pyro/infer/mcmc/api.py

```diff
@@ -310,9 +316,19 @@ def __init__(self, kernel, num_samples, warmup_steps=None, initial_params=None,
 
     def run(self, *args, **kwargs):
         z_acc = defaultdict(lambda: [[] for _ in range(self.num_chains)])
+        divergences = [None for _ in range(self.num_chains)]
         for sample, chain_id in self.sampler.run(*args, **kwargs):
-            for k, v in sample.items():
-                z_acc[k][chain_id].append(v)
+            if isinstance(sample, dict):
+                for k, v in sample.items():
+                    z_acc[k][chain_id].append(v)
+            else:  # diverging info
+                divergences[chain_id] = sample
+
+        diverging_mask = torch.zeros((self.num_chains, self.num_samples), dtype=torch.bool)
+        for i in range(self.num_chains):
+            diverging_mask[i][divergences[i]] = True
+        self._diverging_mask = diverging_mask
```

Comment:
@neerajprad Users can get the mask for diverging samples here. Could you suggest me a pattern to expose this information for user? I think we can make this a property or create a method returning this value. I would like to propose the following pattern:

Make this MCMC class have some methods as follows
+ `run` returns `self` as in the old API
+ `marginal` returns samples as in the old API
+ `diagnostics` returns diagnostics as in the old API with additional `diverging` (or better name) field
+ `summary` prints summary table (which I'll address in #1934)

Reply:
I largely agree with your suggestions. My thoughts are as follows:
 - Let us not have `.run` return anything. I think if a function / method both sets states and returns something, it usually causes confusion.
 - `.marginal(flatten=True)` can return samples where the chain dim is flattened out by default.
 -  I think `diagnostics`, and `summary` will be great to add as well. 
 - Also, let us consider renaming `kernel._diagnostics` to `kernel._logging` or something, and `kernel.post_sampling()` to `kernel.diagnostics()` since diagnostics usually refers to post sampling convergence related diagnostics.